crypto/x509/x509_cmp.c

```diff
@@ -214,13 +214,15 @@ int X509_add_cert(STACK_OF(X509) *sk, X509 *cert, int flags)
         if (ret != 0)
             return ret > 0 ? 1 : 0;
     }
+
+    if ((flags & X509_ADD_FLAG_UP_REF) != 0 && !X509_up_ref(cert))
+        return 0;
     if (!sk_X509_insert(sk, cert,
                         (flags & X509_ADD_FLAG_PREPEND) != 0 ? 0 : -1)) {
+        X509_free(cert);
```

Comment:
Note that I moved the `X509_up_ref(cert)` call before the `sk_X509_insert()` because it can be undone (using `X509_free(cert)`) more easily in case the `sk_X509_insert()` failed than undoing the `sk_X509_insert()` in case the `X509_up_ref(cert)` failed.
The problem with the original implementation was that it simply ignored the case that `(void)X509_up_ref(cert)` failed, which could lead to a double free.

I generally agree a separate PR would be more clean, but I wanted to avoid setting up many new PRs,
and now we've already started discussing these fixes here in this PR.
This change is in a separate commit (and likewise the related fix in `cms_lib.c`).

Reply:
Ah, now I got what you meant: the `X509_free()` itself is not wrong, 
but it needs to be guarded by `if ((flags & X509_ADD_FLAG_UP_REF) != 0)` - done now.

Reply:
It is weird that `CRYPTO_UP_REF()` and misleading always returns 1 - should better have type `void`.
And having a look at 
```
int X509_up_ref(X509 *x)
{
    int i;

    if (CRYPTO_UP_REF(&x->references, &i) <= 0)
        return 0;

    REF_PRINT_COUNT("X509", x);
    REF_ASSERT_ISNT(i < 2);
    return i > 1;
}
```
reveals that there may be other reasons for it it fail.
So I still believe precaution should be taken by its callers.

Reply:
There are no good reasons for `CRYPTO_UP_REF` to fail. If it ever does, nothing in the application state can be trusted from that point on, you've lost memory safety. The primitive needs to be reliable.

I am not impressed by pretence of possible, later recoverable, transient problems atomically incrementing a refcount. It needs to **just work**™.

Reply:
I cannot judge whether `CRYPTO_UP_REF()` might fail or not - my complaint above was simply that the mere fact that it returns a value is misleading, assuming what you wrote before, namely that it cannot fail.

My main point was that `X509_up_ref()` apparently can fail due to other effects (also related somehow to `CRYPTO_UP_REF()`), at least according to its implementation quoted above. If this is impression of mine is wrong, then also the implementation of `X509_up_ref()` is very misleading and could be reduced to:
```
{
    int dummy;

    (void)CRYPTO_UP_REF(&x->references, &dummy);
    return 1;
}
```
and then in turn it would be misleading that `X509_up_ref()` has a return value at all.